main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala

```diff
@@ -273,9 +274,10 @@ object CodeGen extends Phase[ExecutableAst.Root, ExecutableAst.Root] {
                                 functions: List[ExecutableAst.Def],
                                 declarations: Map[Symbol.DefnSym, Type],
                                 interfaces: Map[Type, FlixClassName],
-                                enums: Map[(Type, String), (QualName, ExecutableAst.Case)],
+                                enums: Map[(Type, String), (ExecutableAst.Case, Type)],
                                 visitor: MethodVisitor,
-                                entryPoint: Label)(expr: Expression): Unit = expr match {
+                                entryPoint: Label,
+                                options: Options)(expr: Expression): Unit = expr match {
```

Comment:
This signature is getting really long. I think it can be collapsed like so:

- The functions are available in the AST root?
- The declarations are also available in the AST root?
- The interfaces and enums looks like something that can be grouped into one algebraic data type.
- The visitor is necessary.
- The entryPoint is necessary.
- The options is necessary (but might as well be accessed from the Flix object).

In other Flix phases we are starting to use this pattern (basically just nested functions):

```
def whaterver(root: Root)(implicit flix: Flix): .. {

  def doStuffWithExp(exp: Exp, ...) {
   // notice that in here both the AST root and the Flix object is available.
  }
}
```

This structure nicely "emulates" a form of global variables in a phase, but in a purely functional way using lexical scope.

Reply:
Yes, functions and declarations are available in AST root, even enums and interfaces are from root. The problem is, you need to traverse the whole tree to extract these informations. We can talk about these later.